apps/api/src/notifications/notifications.service.ts

```diff
@@ -0,0 +1,314 @@
+          }
+      } catch (e) {
+        Logger.error(e)
+        throw new BadRequestException('Failed to get campaign info')
+      }
+
+    // Add email to general marketing notifications list
+    const mainList = this.config.get('sendgrid.marketingListId')
+    mainList && listIds.push(mainList)
+
+    // Add to marketing platform
+    try {
+      await this.marketingNotificationsProvider.addContactsToList({
+        contacts: [contact],
+        list_ids: listIds,
+      })
+    } catch (e) {
+      Logger.error('Failed to subscribe email', e)
+      throw new BadRequestException('Failed to subscribe email')
+    }
+
+    // If no prior consent has been given by a registered user
+    if (registered && !registered.newsletter)
+      try {
+        await this.personService.update(registered.id, { newsletter: true })
+      } catch (e) {
+        Logger.error('Failed to update user consent', e)
+        throw new BadRequestException('Failed to update user consent')
+      }
+    // If the email is not registered - create/update a saparate notification consent record
+    else if (!registered)
+      try {
+        await this.personService.updateUnregisteredNotificationConsent(data.email, data.consent)
+      } catch (e) {
+        Logger.error('Failed to save unregistered consent', e)
+        throw new BadRequestException('Failed to save consent')
+      }
+
+    return { email: data.email, subscribed: true }
+  }
+
+  async subscribe(user: KeycloakTokenParsed) {
+    let userInfo: Person | null
+    try {
+      userInfo = await this.personService.findOneByKeycloakId(user.sub)
+    } catch (e) {
+      return new BadRequestException('Failed to get user')
+    }
+
+    if (!userInfo) return new BadRequestException('User not found')
+
+    const contact: SendGridParams['ContactData'] = {
+      email: userInfo.email,
+      first_name: userInfo?.firstName || '',
+      last_name: userInfo?.lastName || '',
+    }
+
+    const listIds: string[] = []
+
+    // Add email to general marketing notifications list
+    const mainList = this.config.get('sendgrid.marketingListId')
+    mainList && listIds.push(mainList)
+
+    // Add to marketing platform
+    try {
+      await this.marketingNotificationsProvider.addContactsToList({
+        contacts: [contact],
+        list_ids: listIds,
+      })
+    } catch (e) {
+      Logger.error('Failed to subscribe email', e)
+      throw new BadRequestException('Failed to subscribe email')
+    }
+
+    if (!userInfo.newsletter)
+      try {
+        await this.personService.update(userInfo.id, { newsletter: true })
+      } catch (e) {
+        Logger.error('Failed to update user consent', e)
+        throw new BadRequestException('Failed to update user consent')
+      }
+
+    return { email: userInfo.email, subscribed: true }
+  }
+
+  generateHash(): string {
+    const hash = crypto
+      .randomBytes(Math.ceil(32 / 2))
+      .toString('hex')
+      .slice(0, 32)
+
+    return hash
+  }
```

Comment:
here instead of randomBytes, can we use the person.id which is unique already?
This way there will be no need to add a new field mailHash in the database, just when the link needs to be checked, we can generate it again from the person.id to compare.

Also let's not slice, but take the full string.

To illustrate what I mean in code:
crypto.createHash('sha256').update(person.id).digest('hex')

Reply:
I'm using the id for the respective document now